firmware/kivipallur_firmware_3.0/Core/Src/main.c

```diff
@@ -0,0 +1,1231 @@
+
+		int16_t position;
+		position = (int16_t)TIM1->CNT;
+		int32_t pwmValue = PIDcontrol(&motor1Control, position);
+		HAL_GPIO_WritePin(MOT1_PWM_2_GPIO_Port, MOT1_PWM_2_Pin, (pwmValue < 0) ? GPIO_PIN_SET : GPIO_PIN_RESET);
+		//wake_drivers_up();
+		TIM1->CCR2 = (uint16_t)clamp((pwmValue < 0) ? -pwmValue : pwmValue, 0, 65535);
+
+		position = (int16_t)TIM2->CNT;
+		pwmValue = PIDcontrol(&motor2Control, position);
+		HAL_GPIO_WritePin(MOT2_PWM_2_GPIO_Port, MOT2_PWM_2_Pin, (pwmValue < 0) ? GPIO_PIN_SET : GPIO_PIN_RESET);
+		//wake_drivers_up();
+		TIM2->CCR2 = (uint16_t)clamp((pwmValue < 0) ? -pwmValue : pwmValue, 0, 65535);
+
+		position = (int16_t)TIM3->CNT;
+		pwmValue = PIDcontrol(&motor3Control, position);
+		HAL_GPIO_WritePin(MOT3_PWM_2_GPIO_Port, MOT3_PWM_2_Pin, (pwmValue < 0) ? GPIO_PIN_SET : GPIO_PIN_RESET);
+		//wake_drivers_up();
+		TIM2->CCR4 = (uint16_t)clamp((pwmValue < 0) ? -pwmValue : pwmValue, 0, 65535);
+
+		timer+=1;
+		if (timer == 50){
+			timer = 0;
+			enable_pid = 0;
+			motor1Control.speed = 0;
+			TIM1->CCR1 = 0;
+			TIM1->CCR2 = 0;
+			motor2Control.speed = 0;
+			TIM2->CCR1 = 0;
+			TIM2->CCR2 = 0;
+			motor3Control.speed = 0;
+			TIM2->CCR3 = 0;
+			TIM2->CCR4 = 0;
+			Thrower_Send(0);
+		}
+	}
+}
+
+void wake_drivers_up() {
+	HAL_GPIO_WritePin(MOT_SLEEP_GPIO_Port, MOT_SLEEP_Pin, 0);
+		for (int i = 0; i<300; i++) {
+			__asm("nop");
+	  	  }
+	HAL_GPIO_WritePin(MOT_SLEEP_GPIO_Port, MOT_SLEEP_Pin, 1);
+}
+/* USER CODE END 0 */
+
+/**
+  * @brief  The application entry point.
+  * @retval int
+  */
+int main(void)
+{
+  /* USER CODE BEGIN 1 */
+
+  /* USER CODE END 1 */
+
+  /* MCU Configuration--------------------------------------------------------*/
+
+  /* Reset of all peripherals, Initializes the Flash interface and the Systick. */
+  HAL_Init();
+
+  /* USER CODE BEGIN Init */
+
+  /* USER CODE END Init */
+
+  /* Configure the system clock */
+  SystemClock_Config();
+
+  /* USER CODE BEGIN SysInit */
+
+  /* USER CODE END SysInit */
+
+  /* Initialize all configured peripherals */
+  MX_GPIO_Init();
+  MX_TIM1_Init();
+  MX_RTC_Init();
+  MX_TIM2_Init();
+  MX_TIM3_Init();
+  MX_TIM8_Init();
+  MX_TIM17_Init();
+  MX_TIM16_Init();
+  MX_TIM4_Init();
+  MX_TIM15_Init();
+  MX_USB_Device_Init();
+  MX_DMA_Init();
+  MX_ADC1_Init();
+  MX_CRC_Init();
+  MX_TIM6_Init();
+  /* USER CODE BEGIN 2 */
+
+  HAL_TIM_PWM_Start(&htim15, TIM_CHANNEL_2);
+  HAL_TIM_PWM_Start(&htim1, TIM_CHANNEL_1);
+  HAL_TIM_PWM_Start(&htim1, TIM_CHANNEL_2);
+  HAL_TIM_PWM_Start(&htim2, TIM_CHANNEL_1);
+  HAL_TIM_PWM_Start(&htim2, TIM_CHANNEL_2);
+  HAL_TIM_PWM_Start(&htim2, TIM_CHANNEL_3);
+  HAL_TIM_PWM_Start(&htim2, TIM_CHANNEL_4);
+
+  HAL_TIM_Encoder_Start(&htim3, TIM_CHANNEL_1);
+  HAL_TIM_Encoder_Start(&htim3, TIM_CHANNEL_2);
+  HAL_TIM_Encoder_Start(&htim4, TIM_CHANNEL_1);
+  HAL_TIM_Encoder_Start(&htim4, TIM_CHANNEL_2);
+  HAL_TIM_Encoder_Start(&htim8, TIM_CHANNEL_1);
+  HAL_TIM_Encoder_Start(&htim8, TIM_CHANNEL_2);
```

Comment:
Encoder mode timers can (or maybe should) be started with a single line.

For example:
`HAL_TIM_Encoder_Start(&htim1, TIM_CHANNEL_1 | TIM_CHANNEL_2);`

https://ut-robotics.github.io/picr22-home/basketball_robot_guide/electronics/firmware.html#_timer_code